File: 0844-backspace-string-compare/0844-backspace-string-compare.py
class Solution(object):
    def backspaceCompare(self, s, t):
       i = len(s) - 1
       j = len(t) - 1
       while i >= 0 or j >= 0:

        skips=0
        while i >= 0:
            if s[i] == '#':
                skips += 1
                i -= 1
            elif skips > 0:
                skips -= 1
                i -= 1
            else:
                break

        skipt =0
        while j >= 0:
            if t[j] == '#':
                skipt += 1
                j -= 1 
            elif skipt > 0:
                skipt -= 1
                j -= 1
            else:
                break

        if i >= 0 and j >= 0:
            if s[i] != t[j]:
              return False
        elif i >= 0 or j >= 0:
            return False

        i -= 1
        j -= 1
       return True
        

File: 0844-backspace-string-compare/0844-backspace-string-compare.cpp
class Solution {
public:
    bool backspaceCompare(string s, string t) {
        int i = s.length() - 1;
        int j = t.length() - 1;

        while (i >= 0 || j >= 0) {

            // Find next valid character in s
            int skipS = 0;
            while (i >= 0) {
                if (s[i] == '#') {
                    skipS++;
                    i--;
                }
                else if (skipS > 0) {
                    skipS--;
                    i--;
                }
                else {
                    break;
                }
            }

            // Find next valid character in t
            int skipT = 0;
            while (j >= 0) {
                if (t[j] == '#') {
                    skipT++;
                    j--;
                }
                else if (skipT > 0) {
                    skipT--;
                    j--;
                }
                else {
                    break;
                }
            }

            // Compare current valid characters
            if (i >= 0 && j >= 0) {
                if (s[i] != t[j])
                    return false;
            }
            else if (i >= 0 || j >= 0) {
                return false;
            }

            i--;
            j--;
        }

        return true;
    }
};